File: src/main/java/main/ui/renderer/CombListCellRenderer.java
package main.ui.renderer;

import main.App;
import main.puzzle.Board;
import main.puzzle.assembly.ChipFreq;
import main.util.Fn;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.*;

/**
 *
 * @author Bunnyspa
 */
public class CombListCellRenderer extends DefaultListCellRenderer {

    private final App app;
    private final JList<?> combChipFreqList;

    public CombListCellRenderer(App app, JList<?> combChipFreqList) {
        super();
        this.app = app;
        this.combChipFreqList = combChipFreqList;
    }

    @NotNull
    @Override
    public Component getListCellRendererComponent(@NotNull JList<?> list, Object value, int index, boolean isSelected, boolean cellHasFocus) {
        DefaultListCellRenderer cr = (DefaultListCellRenderer) super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus);
        try {
            cr.setHorizontalAlignment(CENTER);
            Board b = (Board) value;
            cr.setText(Fn.fPercStr(b.getStatPerc()));

            boolean freqChipSelected = !combChipFreqList.isSelectionEmpty();
            boolean freqChipIncluded = false;
            if (!combChipFreqList.isSelectionEmpty()) {
                String freqID = ((ChipFreq) combChipFreqList.getSelectedValue()).chip.getID();
                freqChipIncluded = b.getChipIDs().stream().anyMatch((id) -> id.equals(freqID));
            }

            DefaultListModel<?> combList = (DefaultListModel<?>) list.getModel();
            Board bMax = (Board) combList.firstElement();
            Board bMin = (Board) combList.lastElement();
            cr.setForeground(freqChipSelected
                    ? (isSelected ? Color.GRAY : Color.BLACK)
                    : (isSelected ? Color.WHITE : Color.BLACK));
            cr.setBackground(freqChipIncluded ? Color.LIGHT_GRAY
                    : freqChipSelected ? Color.WHITE
                            : Fn.percColor(app.orange(), app.green(), app.blue(), b.getStatPerc(), bMin.getStatPerc(), bMax.getStatPerc()));
        } catch (Exception ignored) {
        }
        return cr;
    }
}

File: src/main/java/main/ui/renderer/CombListCellRenderer.kt
package main.ui.renderer

import main.App
import main.puzzle.Board
import main.puzzle.assembly.ChipFreq
import main.util.Fn
import java.awt.Color
import java.awt.Component
import javax.swing.DefaultListCellRenderer
import javax.swing.DefaultListModel
import javax.swing.JList

/**
 *
 * @author Bunnyspa
 */
class CombListCellRenderer(private val app: App, private val combChipFreqList: JList<*>?) : DefaultListCellRenderer() {
    override fun getListCellRendererComponent(
        list: JList<*>,
        value: Any,
        index: Int,
        isSelected: Boolean,
        cellHasFocus: Boolean
    ): Component {
        val cr =
            super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus) as DefaultListCellRenderer
        try {
            cr.horizontalAlignment = CENTER
            val b = value as Board
            cr.text = Fn.fPercStr(b.getStatPerc())
            val freqChipSelected = !combChipFreqList!!.isSelectionEmpty
            var freqChipIncluded = false
            if (!combChipFreqList.isSelectionEmpty) {
                val freqID = (combChipFreqList.selectedValue as ChipFreq).chip!!.id
                freqChipIncluded = b.getChipIDs().stream().anyMatch { id: String? -> id == freqID }
            }
            val combList = list.model as DefaultListModel<*>
            val bMax = combList.firstElement() as Board
            val bMin = combList.lastElement() as Board
            cr.foreground =
                if (freqChipSelected) if (isSelected) Color.GRAY else Color.BLACK else if (isSelected) Color.WHITE else Color.BLACK
            cr.background =
                if (freqChipIncluded) Color.LIGHT_GRAY else if (freqChipSelected) Color.WHITE else Fn.percColor(
                    app.orange()!!, app.green()!!, app.blue()!!, b.getStatPerc(), bMin.getStatPerc(), bMax.getStatPerc()
                )
        } catch (ignored: Exception) {
        }
        return cr
    }
}